From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sat, 1 Feb 2025 19:50:02 +0300
Subject: [PATCH] SparklyPaper: Skip distanceToSqr call in
 ServerEntity#sendChanges if the delta movement hasn't changed

Original project: https://github.com/SparklyPower/SparklyPaper

Patch description:

The "distanceToSqr" call is a bit expensive, so avoiding it is pretty nice, around ~15% calls are skipped with this check
We could also check if the x,y,z coordinates are equal, but for now, let's just keep the identity check, which also helps us since Minecraft's code does reuse the original delta movement Vec3 object

diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 60af82925d2f7e605bf7340cf01f1a5668f4d725..4b4dd843e00f26c85ba9cfed388b1adadf247bcf 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -202,23 +202,27 @@ public class ServerEntity {
 
                 if (this.entity.needsSync || this.trackDelta || this.entity instanceof LivingEntity livingEntity && livingEntity.isFallFlying()) {
                     Vec3 movement = this.entity.getDeltaMovement();
-                    double diff = movement.distanceToSqr(this.lastSentMovement);
-                    if (diff > 1.0E-7 || diff > 0.0 && movement.lengthSqr() == 0.0) {
-                        this.lastSentMovement = movement;
-                        if (this.entity instanceof AbstractHurtingProjectile projectile) {
-                            this.synchronizer
-                                .sendToTrackingPlayers(
-                                    new ClientboundBundlePacket(
-                                        List.of(
-                                            new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement),
-                                            new ClientboundProjectilePowerPacket(projectile.getId(), projectile.accelerationPower)
+                    // DivineMC start - Skip "distanceToSqr" call in "ServerEntity#sendChanges" if the delta movement hasn't changed
+                    if (movement != this.lastSentMovement) {
+                        double diff = movement.distanceToSqr(this.lastSentMovement);
+                        if (diff > 1.0E-7 || diff > 0.0 && movement.lengthSqr() == 0.0) {
+                            this.lastSentMovement = movement;
+                            if (this.entity instanceof AbstractHurtingProjectile projectile) {
+                                this.synchronizer
+                                    .sendToTrackingPlayers(
+                                        new ClientboundBundlePacket(
+                                            List.of(
+                                                new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement),
+                                                new ClientboundProjectilePowerPacket(projectile.getId(), projectile.accelerationPower)
+                                            )
                                         )
-                                    )
-                                );
-                        } else {
-                            this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
+                                    );
+                            } else {
+                                this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
+                            }
                         }
                     }
+                    // DivineMC end - Skip "distanceToSqr" call in "ServerEntity#sendChanges" if the delta movement hasn't changed
                 }
 
                 if (packet != null) {
